File: java/limax/src/limax/node/js/modules/Url.java
package limax.node.js.modules;

import java.net.URI;
import java.util.HashMap;
import java.util.Map;

import limax.node.js.EventLoop;
import limax.node.js.Module;

public final class Url implements Module {
	public Url(EventLoop eventLoop) {
	}

	public Map<String, Object> parse(String url) throws Exception {
		URI uri = new URI(url).normalize();
		String scheme = uri.getScheme();
		if (scheme != null)
			scheme = scheme.toLowerCase();
		String userInfo = uri.getUserInfo();
		String host = uri.getHost();
		if (host != null)
			host = host.toLowerCase();
		int port = uri.getPort();
		String path = uri.getPath();
		if (path.isEmpty())
			path = "/";
		String query = uri.getQuery();
		String fragment = uri.getFragment();
		uri = new URI(scheme, userInfo, host, port, path, query, fragment);
		Map<String, Object> map = new HashMap<>();
		map.put("href", uri.toString());
		map.put("protocol", scheme != null ? scheme + ":" : null);
		map.put("slashes", uri.getAuthority() != null);
		map.put("host", port != -1 ? host + ":" + port : host);
		map.put("auth", userInfo);
		map.put("hostname", host);
		map.put("port", port != -1 ? port : null);
		map.put("pathname", path);
		map.put("search", query != null ? "?" + query : null);
		map.put("path", path + (query != null ? "?" + query : ""));
		map.put("query", query);
		map.put("hash", fragment != null ? "#" + fragment : null);
		return map;
	}

	public String resolve(String from, String to) throws Exception {
		return new URI(from).resolve(to).toString();
	}
}

File: java/limax/src/limax/node/js/modules/Url.js
var querystring = require("querystring");
exports.format = function(urlObject) {
	if (typeof(uriObject) === 'string')
		urlObject = this.parse(urlObject);
	var r = '';
	if (typeof(urlObject.protocol) === 'string')
		r += protocol;
	else if (urlObject.protocol && typeof(urlObject.protocol) !== 'string')
		throw new Error();
	if (r.length > 0 && r.charAt(r.length - 1) != ':')
		r += ':';
	if (urlObject.slashes === true || r.startsWith('http:') || r.startsWith('https:') || r.startsWith('ftp:') || r.startsWith('gopher:') || r.startsWith('file:'))
		r += '//';
	if (urlObject.auth && (urlObject.host || urlObject.hostname))
		r += urlObject.auth + '@';
	if (!urlObject.host) {
		if (typeof(urlObject.hostname) === 'string')
			r += urlObject.hostname;
		else if (!urlObject.hostname && typeof(urlObject.hostname) !== 'string')
			throw new Error();
		if (urlObject.port && !urlObject.hostname)
			r += ':' + urlObject.port;
	} else {
		r += urlObject.host;
	}
	if (typeof(urlObject.pathname) === 'string' && urlObject.pathname.length > 0) {
		if (!urlObject.pathname.startsWith('/'))
			r += '/';
		r += urlObject.pathname;
	} else if (urlObject.pathname && typeof(urlObject.pathname) !== 'string')
		throw new Error();
	if (!urlObject.search && typeof(urlObject.query) === 'object' && urlObject.query)
		r += '?' + querystring.stringify(urlObject.query);
	else if (typeof(urlObject.search) === 'string') {
		if (!urlObject.search.startsWith('?'))
			r += '?';
		r += urlObject.search;
	} else if(urlObject.search && typeof(urlObject.search) !== 'string')
		throw new Error();
	if (typeof(urlObject.hash) === 'string') {
		if (!urlObject.hash.startsWith('#'))
			r += '#';
		r += urlObject.hash;
	} else if (urlObject.hash && typeof(urlObject.hash) !== 'string')
		throw new Error();
	return r;
}
exports.parse = function(urlString, parseQueryString) {
	var r = {};
	java.parse(urlString).forEach(function(k, v) {r[k] = v});
	if (parseQueryString === true && r.query)
		r.query = querystring.parse(r.query);
	return r;
}
exports.resolve = function(from, to) {
	return java.resolve(from, to);
}
